rllib/utils/explorations/epsilon_greedy.py

```diff
@@ -0,0 +1,177 @@
+from ray.rllib.utils.explorations.exploration import Exploration
+from ray.rllib.utils.framework import try_import_tf, try_import_torch, \
+    get_variable
+from ray.rllib.utils.schedules import PiecewiseSchedule
+
+tf = try_import_tf()
+torch, _ = try_import_torch()
+
+
+class EpsilonGreedy(Exploration):
+    """
+    An epsilon-greedy Exploration class that produces exploration actions
+    when given a Model's output and a current epsilon value (based on some
+    Schedule).
+    """
+
+    def __init__(self,
+                 action_space,
+                 initial_epsilon=1.0,
+                 final_epsilon=0.1,
+                 epsilon_timesteps=int(1e5),
+                 worker_info=None,
+                 epsilon_schedule=None,
+                 framework="tf"):
+        """
+        Args:
+            action_space (Space): The gym action space used by the environment.
+            initial_epsilon (float): The initial epsilon value to use.
+            final_epsilon (float): The final epsilon value to use.
+            epsilon_timesteps (int): The time step after which epsilon should
+                always be `final_epsilon`.
+            worker_info (Optional[Dict[str,any]]): Dict with keys:
+                `num_workers`: The overall number of workers used.
+                `worker_index`: The index of the Worker using this Exploration.
+            epsilon_schedule (Optional[Schedule]): An optional Schedule object
+                to use (instead of constructing one from the given parameters).
+            framework (Optional[str]): One of None, "tf", "torch".
+        """
+        # For now, require Discrete action space (may loosen this restriction
+        # in the future).
+        assert isinstance(action_space, gym.spaces.Discrete)
+        assert framework is not None
+        super().__init__(
+            action_space=action_space,
+            worker_info=worker_info,
+            framework=framework)
+
+        self.epsilon_schedule = epsilon_schedule or PiecewiseSchedule(
+            endpoints=[(0, initial_epsilon), (epsilon_timesteps,
+                                              final_epsilon)],
+            outside_value=final_epsilon,
+            framework=self.framework)
+
+        # The current time_step value (tf-var or python int).
+        self.last_time_step = get_variable(
+            0, framework=framework, tf_name="time-step")
+
+    @override(Exploration)
+    def get_exploration_action(self,
+                               action,
+                               model=None,
+                               action_dist=None,
+                               explore=True,
+                               time_step=None):
+        # TODO(sven): This is hardcoded. Put a meaningful error, in case model
+        # API is not as required.
+        q_values = model.q_value_head(model._last_output)[0]
+        if self.framework == "tf":
+            return self._get_tf_exploration_action_op(action, explore,
```

Comment:
Can we have this for both tf and torch? Weird that the main body is torch below.

Reply:
Sure. Came from the fact that we had it like that for pure python (main body). Then I added torch.

Reply:
done